File: src/invoicing/invoicing-service.ts
import { createClient, SupabaseClient } from '@supabase/supabase-js';
import { InvoicePdfGenerator, InvoiceData } from './pdf-generator';
import { InvoicePdfStorage } from './supabase-storage';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

/**
 * Complete invoicing service
 * Handles: database queries, PDF generation, storage, and metadata tracking
 */
export class InvoicingService {
  private supabase: SupabaseClient;
  private pdfGenerator: InvoicePdfGenerator;
  private pdfStorage: InvoicePdfStorage;
  private tempDir: string;

  constructor(
    supabaseUrl: string,
    supabaseKey: string,
    templatePath?: string
  ) {
    this.supabase = createClient(supabaseUrl, supabaseKey);
    this.pdfGenerator = new InvoicePdfGenerator(templatePath);
    this.pdfStorage = new InvoicePdfStorage(supabaseUrl, supabaseKey);
    this.tempDir = path.join(os.tmpdir(), 'invoice-pdfs');

    // Create temp directory if it doesn't exist
    if (!fs.existsSync(this.tempDir)) {
      fs.mkdirSync(this.tempDir, { recursive: true });
    }
  }

  /**
   * Initialize service (create storage bucket, setup browser)
   */
  async initialize() {
    await this.pdfStorage.ensureBucket();
    await this.pdfGenerator.initialize();
  }

  /**
   * Fetch invoice data from database
   */
  async fetchInvoiceData(invoiceId: string): Promise<InvoiceData | null> {
    try {
      // Fetch invoice from public schema
      const { data: invoiceData, error: invoiceError } = await this.supabase
        .from('invoices')
        .select('*')
        .eq('id', invoiceId)
        .single();

      if (invoiceError) {
        console.error('Error fetching invoice:', invoiceError);
        return null;
      }

      // Fetch entity and contact details
      const { data: entityData } = await this.supabase
        .from('entities')
        .select('*')
        .eq('id', invoiceData.entity_id)
        .single();

      const { data: contactData } = await this.supabase
        .from('contacts')
        .select('*')
        .eq('id', invoiceData.contact_id)
        .single();

      // Map database record to InvoiceData interface
      const mapped: InvoiceData = {
        id: invoiceData.id,
        entity_id: invoiceData.entity_id,
        contact_id: invoiceData.contact_id,
        invoice_number: invoiceData.invoice_number,
        purchase_order_number: invoiceData.purchase_order_number,
        invoice_date: invoiceData.invoice_date,
        due_date: invoiceData.due_date,
        project: invoiceData.project,
        project_id: invoiceData.project_id,
        subtotal_amount: parseFloat(invoiceData.subtotal_amount || '0'),
        gst_amount: parseFloat(invoiceData.gst_amount || '0'),
        total_amount: parseFloat(invoiceData.total_amount || '0'),
        status: invoiceData.status,
        client_name: contactData?.name || invoiceData.client || '',
        client_email: contactData?.email,
        client_abn: contactData?.abn,
        billing_address: invoiceData.billed_to,
        entity_name: entityData?.name || '',
        entity_abn: entityData?.abn || '',
        trading_name: entityData?.trading_name,
        invoice_prefix: invoiceData.invoice_number?.split('-')[0] || 'INV',
        items: [], // No line items in current schema
      };

      return mapped;
    } catch (error) {
      console.error('Error fetching invoice data:', error);
      return null;
    }
  }

  /**
   * Fetch all invoices for an entity
   */
  async fetchEntityInvoices(entityId: string): Promise<InvoiceData[]> {
    try {
      const { data: invoices, error } = await this.supabase
        .from('invoices')
        .select('*')
        .eq('entity_id', entityId)
        .order('invoice_date', { ascending: false });

      if (error) {
        console.error('Error fetching entity invoices:', error);
        return [];
      }

      // Fetch entity and contacts data for enrichment
      const { data: entityData } = await this.supabase
        .from('entities')
        .select('*')
        .eq('id', entityId)
        .single();

      const contactIds = [...new Set((invoices || []).map(i => i.contact_id))];
      const { data: contactsData } = await this.supabase
        .from('contacts')
        .select('*')
        .in('id', contactIds);

      const contactMap = new Map(contactsData?.map(c => [c.id, c]) || []);

      // Map all invoices
      const mapped = (invoices || []).map(inv => ({
        id: inv.id,
        entity_id: inv.entity_id,
        contact_id: inv.contact_id,
        invoice_number: inv.invoice_number,
        purchase_order_number: inv.purchase_order_number,
        invoice_date: inv.invoice_date,
        due_date: inv.due_date,
        project: inv.project,
        project_id: inv.project_id,
        subtotal_amount: parseFloat(inv.subtotal_amount || '0'),
        gst_amount: parseFloat(inv.gst_amount || '0'),
        total_amount: parseFloat(inv.total_amount || '0'),
        status: inv.status,
        client_name: contactMap.get(inv.contact_id)?.name || inv.client || '',
        client_email: contactMap.get(inv.contact_id)?.email,
        client_abn: contactMap.get(inv.contact_id)?.abn,
        billing_address: inv.billed_to,
        entity_name: entityData?.name || '',
        entity_abn: entityData?.abn || '',
        trading_name: entityData?.trading_name,
        invoice_prefix: inv.invoice_number?.split('-')[0] || 'INV',
        items: [], // No line items in current schema
      } as InvoiceData));

      return mapped;
    } catch (error) {
      console.error('Error fetching entity invoices:', error);
      return [];
    }
  }

  /**
   * Generate and store invoice PDF
   * Returns URL and metadata for database insertion
   */
  async generateInvoicePdf(invoiceId: string): Promise<{
    pdf_url: string;
    file_name: string;
    storage_path: string;
    pdf_size_bytes: number;
  } | null> {
    try {
      // Fetch invoice data
      const invoiceData = await this.fetchInvoiceData(invoiceId);
      if (!invoiceData) {
        throw new Error(`Invoice not found: ${invoiceId}`);
      }

      // Generate PDF locally
      const tempPdfPath = path.join(
        this.tempDir,
        `${invoiceData.invoice_number.replace(/\//g, '-')}-${Date.now()}`
      );

      const pdfPath = await this.pdfGenerator.generatePdf(
        invoiceData,
        tempPdfPath
      );

      // Get file size
      const stats = fs.statSync(pdfPath);
      const fileSizeBytes = stats.size;

      // Upload to Supabase Storage
      const publicUrl = await this.pdfStorage.uploadPdf(
        pdfPath,
        invoiceData.invoice_number,
        invoiceData.entity_id
      );

      // Generate storage path (same as used in storage)
      const date = new Date();
      const year = date.getFullYear();
      const storagePath = `${invoiceData.entity_id}/${year}/${invoiceData.invoice_number.replace(/\//g, '-')}.pdf`;

      // Generate file name
      const fileName = `${invoiceData.invoice_number}.pdf`;

      // Clean up temporary file
      fs.unlinkSync(pdfPath);

      console.log(`✓ Successfully generated invoice PDF for ${invoiceData.invoice_number}`);

      return {
        pdf_url: publicUrl,
        file_name: fileName,
        storage_path: storagePath,
        pdf_size_bytes: fileSizeBytes,
      };
    } catch (error) {
      console.error('Error generating invoice PDF:', error);
      return null;
    }
  }

  /**
   * Save PDF metadata to database
   * Note: Currently skipped as invoice_pdfs table doesn't exist in schema
   * PDF URL is already stored in Supabase Storage and accessible
   */
  async savePdfMetadata(
    invoiceId: string,
    pdfData: {
      pdf_url: string;
      file_name: string;
      storage_path: string;
      pdf_size_bytes: number;
    }
  ) {
    try {
      // TODO: When invoice_pdfs table is created, implement metadata storage here
      // const { error } = await this.supabase
      //   .from('invoice_pdfs')
      //   .insert({
      //     invoice_id: invoiceId,
      //     pdf_url: pdfData.pdf_url,
      //     file_name: pdfData.file_name,
      //     storage_path: pdfData.storage_path,
      //     pdf_size_bytes: pdfData.pdf_size_bytes,
      //     generated_at: new Date().toISOString(),
      //   });

      console.log(`✓ PDF stored for invoice ${invoiceId}: ${pdfData.pdf_url}`);
    } catch (error) {
      console.error('Error saving PDF metadata:', error);
      throw error;
    }
  }

  /**
   * Complete workflow: Generate PDF and save metadata
   */
  async generateAndStorePdf(invoiceId: string): Promise<boolean> {
    try {
      const pdfData = await this.generateInvoicePdf(invoiceId);
      if (!pdfData) {
        return false;
      }

      await this.savePdfMetadata(invoiceId, pdfData);
      return true;
    } catch (error) {
      console.error('Error in generate and store workflow:', error);
      return false;
    }
  }

  /**
   * Generate PDFs for multiple invoices
   */
  async generateBatch(invoiceIds: string[]): Promise<{
    success: number;
    failed: number;
    results: Array<{ invoiceId: string; success: boolean; error?: string }>;
  }> {
    const results = [];
    let successCount = 0;
    let failureCount = 0;

    for (const invoiceId of invoiceIds) {
      try {
        const success = await this.generateAndStorePdf(invoiceId);
        if (success) {
          successCount++;
          results.push({ invoiceId, success: true });
        } else {
          failureCount++;
          results.push({ invoiceId, success: false, error: 'Generation failed' });
        }
      } catch (error) {
        failureCount++;
        results.push({
          invoiceId,
          success: false,
          error: String(error),
        });
      }
    }

    console.log(
      `✓ Batch complete: ${successCount} succeeded, ${failureCount} failed`
    );

    return {
      success: successCount,
      failed: failureCount,
      results,
    };
  }

  /**
   * Clean up resources
   */
  async cleanup() {
    await this.pdfGenerator.close();

    // Clean up temp directory
    if (fs.existsSync(this.tempDir)) {
      const files = fs.readdirSync(this.tempDir);
      for (const file of files) {
        fs.unlinkSync(path.join(this.tempDir, file));
      }
      fs.rmdirSync(this.tempDir);
    }
  }
}

/**
 * Example usage / CLI entry point
 */
async function main() {
  const supabaseUrl = process.env.SUPABASE_URL || '';
  const supabaseKey = process.env.SUPABASE_KEY || '';

  if (!supabaseUrl || !supabaseKey) {
    console.error('Missing SUPABASE_URL or SUPABASE_KEY environment variables');
    process.exit(1);
  }

  const service = new InvoicingService(supabaseUrl, supabaseKey);

  try {
    await service.initialize();
    console.log('✓ Invoicing service initialized');

    // Example: Generate PDF for a specific invoice
    // Uncomment to test:
    // const success = await service.generateAndStorePdf('invoice-uuid-here');
    // console.log('Generation result:', success);

  } catch (error) {
    console.error('Error:', error);
  } finally {
    await service.cleanup();
  }
}

// Uncomment to run as CLI
// if (require.main === module) {
//   main();
// }

File: src/invoicing/invoicing_service.py
"""
Python wrapper for the Invoicing Service that bridges to the TypeScript implementation
and Supabase database operations.
"""

import os
from typing import Optional, List, Dict, Any
from datetime import datetime
import json
from supabase import create_client, Client as SupabaseClient


class InvoicingService:
    """
    Main invoicing service that handles:
    - Invoice data retrieval from Supabase
    - PDF generation and storage
    - Metadata management
    """

    def __init__(self, supabase_url: str, supabase_key: str):
        """Initialize the invoicing service with Supabase credentials."""
        self.supabase_url = supabase_url
        self.supabase_key = supabase_key
        self.supabase: Optional[SupabaseClient] = None

    async def initialize(self) -> None:
        """Initialize Supabase client."""
        try:
            self.supabase = create_client(self.supabase_url, self.supabase_key)
        except Exception as e:
            raise Exception(f"Failed to initialize Supabase client: {str(e)}")

    async def fetchInvoiceData(self, invoice_id: str) -> Optional[Dict[str, Any]]:
        """
        Fetch complete invoice data from the v_all_invoices view (public schema).

        Args:
            invoice_id: UUID of the invoice

        Returns:
            Invoice data dict or None if not found
        """
        if not self.supabase:
            raise Exception("Service not initialized. Call initialize() first.")

        try:
            # Query the public schema view which aggregates all invoice data
            # The Python Supabase client can only access public schema via .table()
            response = self.supabase.table('v_all_invoices') \
                .select('*') \
                .eq('id', invoice_id) \
                .execute()

            if response.data and len(response.data) > 0:
                return response.data[0]

            return None

        except Exception as e:
            print(f"Error fetching invoice {invoice_id}: {str(e)}")
            import traceback
            print(f"Traceback: {traceback.format_exc()}")
            return None

    async def fetchEntityInvoices(
        self,
        entity_id: str,
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """
        Fetch all invoices for a specific entity.

        Args:
            entity_id: UUID of the entity
            limit: Maximum number of invoices to return

        Returns:
            List of invoice dictionaries
        """
        if not self.supabase:
            raise Exception("Service not initialized. Call initialize() first.")

        try:
            # Query the public schema view which aggregates all invoice data
            # Note: The Supabase Python client can handle complex queries,
            # but we need to ensure each method is being called correctly.
            # Using a simpler approach to debug potential issues with ordering/limiting.

            response = self.supabase.table('v_all_invoices') \
                .select('*') \
                .eq('entity_id', entity_id) \
                .execute()

            # Sort and limit in Python if needed
            data = response.data if response.data else []

            # Sort by invoice_date descending
            if data:
                data = sorted(data, key=lambda x: x.get('invoice_date', ''), reverse=True)
                # Limit results
                data = data[:limit]

            return data

        except Exception as e:
            print(f"Error fetching invoices for entity {entity_id}: {str(e)}")
            import traceback
            print(f"Traceback: {traceback.format_exc()}")
            return []

    async def getPdfStatus(self, invoice_id: str) -> Optional[Dict[str, Any]]:
        """
        Check if a PDF has been generated for an invoice.

        Note: PDF checking is not yet implemented - requires Postgres RPC or migration of invoice_pdfs to public schema.

        Args:
            invoice_id: UUID of the invoice

        Returns:
            PDF metadata dict or None if not generated
        """
        if not self.supabase:
            raise Exception("Service not initialized. Call initialize() first.")

        try:
            # TODO: Implement PDF status checking
            # The invoice_pdfs table is in the invoicing schema which Python Supabase client cannot access
            # Options:
            # 1. Create a public view of invoice_pdfs
            # 2. Use a Postgres RPC function for querying
            # 3. Migrate invoice_pdfs table to public schema
            return None

        except Exception as e:
            print(f"Error checking PDF status for {invoice_id}: {str(e)}")
            import traceback
            print(f"Traceback: {traceback.format_exc()}")
            return None

    async def generateAndStorePdf(self, invoice_id: str) -> bool:
        """
        Generate PDF for an invoice and store metadata.
        Note: This is a placeholder that would call the TypeScript implementation.

        Args:
            invoice_id: UUID of the invoice

        Returns:
            True if successful, False otherwise
        """
        try:
            # This would normally call the TypeScript PDF generator
            # For now, we verify the invoice exists and record generation intent
            invoice = await self.fetchInvoiceData(invoice_id)
            if not invoice:
                return False

            # In production, this would:
            # 1. Call TypeScript PDF generator
            # 2. Upload to Supabase Storage
            # 3. Insert metadata

            return True

        except Exception as e:
            print(f"Error generating PDF: {str(e)}")
            return False

    async def generatePdfBatch(
        self,
        invoice_ids: List[str]
    ) -> List[Dict[str, Any]]:
        """
        Generate PDFs for multiple invoices with error handling.

        Args:
            invoice_ids: List of invoice UUIDs

        Returns:
            List of results with success/failure status
        """
        results = []

        for invoice_id in invoice_ids:
            try:
                success = await self.generateAndStorePdf(invoice_id)
                invoice = await self.fetchInvoiceData(invoice_id)

                results.append({
                    "invoice_id": invoice_id,
                    "invoice_number": invoice['invoice_number'] if invoice else None,
                    "success": success
                })
            except Exception as e:
                results.append({
                    "invoice_id": invoice_id,
                    "success": False,
                    "error": str(e)
                })

        return results

    async def cleanup(self) -> None:
        """Clean up resources."""
        # Placeholder for cleanup logic
        pass
